### PR TITLE
Prevent CSRF, Enforce CRUD on Copy Address button for Household object

### PR DESCRIPTION
**Pull request by jjbennett**
# Critical Changes

# Changes
We fixed a security issue with the copying of the address from the older Household object to the related members and from a contact to its related Household object. After you click either "Copy Household Address to Contacts" or the "Copy Address to Household" button, you will now be required to confirm the copying action by clicking the Continue button. Additionally, the Read / Update permission on the Household and Contact address fields is now required to copy the address.

# Issues Closed

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by jjbennett** [comment]
**lurch: attach W-028183

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000007fNRfAAM" target="_blank">W-028183</a>

**Comment by jjbennett** [comment]
@sam-knox - This work item is similar to #3901 in that we will now be enforcing Read/Update permissions and will have a landing page to confirm the action. I wanted to check and see if we should call out the change in the changes section. My hesitation is due to the fact that the majority of our uses now use the Account HH model and this button is only used by the older Household object model. Also attaching screen shots of the page with the corresponding message.
confirmationpage
errorpage

**Comment by jjbennett** [comment]
LOL @seanpat09 - I was admittedly being a bit lazy on this one but you are right and I will get those new tests added. Thanks!

File: src/classes/ADDR_CopyAddrHHObjBTN_CTRL.cls [hunk 1]
```diff
@@ -42,27 +42,53 @@ public with sharing class ADDR_CopyAddrHHObjBTN_CTRL {
...
     /*******************************************************************************************************
     * @description whether to redirect on success, or stay on the page for errors.
     */
-    public boolean redirectPage { get; set; }
+    public Boolean redirectPage { get; set; }
+
+    /*******************************************************************************************************
+    * @description true when the user has permissions required to copy the address
+    */
+    @TestVisible
```
[the remark this answers is not in the document]
Comment (jjbennett) [reply src/classes/ADDR_CopyAddrHHObjBTN_CTRL.cls]: Just a note about the `@TestVisible`, it is there for the private set, it allows us to set the value in a test. Working on creating tests for new methods now. Thanks again.

**Comment by jjbennett** [comment]
@seanpat09 - thanks again for the help! This is ready for review again.

File: src/classes/ADDR_CopyAddrHHObjBTN_TEST.cls [hunk 2]
```diff
@@ -33,84 +33,295 @@
...
+    * @description Verify a note was created for the provided record Id
+    * @param parentId - the id of the record the note belongs to
+    * @return void
+    */
+    private static void verifyNoteRecordCount(Id parentId) {
+        System.assertEquals(1, [SELECT count() FROM Note where ParentID = :parentId]);
     }
+
+    /*******************************************************************************************************************
+    * @description Stub for SObjectDescribe instance
```
Comment (seanpat09) [inline src/classes/ADDR_CopyAddrHHObjBTN_TEST.cls]: Looks like the description from needs to updated from SObjectDescribe to PermissionsMock

**Comment by sam-knox** [comment]
@jjbennett - Go ahead and add a Change note, calling out that it's for the Household object only not the Household Accounts.

**Pull request merged**